pkg/raft/term_cache_test.go

```diff
@@ -0,0 +1,396 @@
+			want: entryID{term: 0, index: 9}.append(5, 6, 7, 8, 8, 8).LogSlice,
+		},
+		{
+			name:      "overwrite test case",
+			cacheSize: 4,
+			// Initial: [t5/10, t6/11, t7/12, t8/13]
+			init: entryID{term: 5, index: 10}.append(6, 7, 8).LogSlice,
+			// Append: [t5/10, t7/11, t8/12, ... t13/17, t13/18, ...]
+			app: entryID{term: 5, index: 10}.append(7, 8, 9, 10, 11, 12, 13, 13, 13, 13, 13, 13).LogSlice,
+			// Test case: [t10/14, t11/15, t12/16, t13/17, t13/18, ...]
+			want: entryID{term: 0, index: 13}.append(10, 11, 12, 13, 13, 13, 13, 13, 13, 13, 13).LogSlice,
+		},
+		{
+			name:      "extreme but possible, overwrite with LogSlice.prev and entry",
+			cacheSize: 4,
+			// Initial: [t5/10, t6/11, t7/12, t8/13]
+			init: entryID{term: 5, index: 10}.append(6, 7, 8).LogSlice,
+			// Append: [t4/9, t10/10]
+			app: entryID{term: 4, index: 9}.append(10).LogSlice,
+			// Test case: [t4/9, t10/10, t10/11, t10/12]
+			want: entryID{term: 0, index: 8}.append(4, 10, 10).LogSlice,
+		},
+		{
+			name:      "overwrite with just one entry",
+			cacheSize: 4,
+			// Initial: [t5/10, t6/11, t7/12, t8/13]
+			init: entryID{term: 5, index: 10}.append(6, 7, 8).LogSlice,
+			// Append: [t5/10, t10/11]
+			app: entryID{term: 5, index: 10}.append(10).LogSlice,
+			// Test case: [t5/10, t10/11, t10/12, t10/13]
+			want: entryID{term: 0, index: 9}.append(5, 10, 10, 10).LogSlice,
+		},
+		{
+			name:      "multiple gaps in between, test term",
+			cacheSize: 4,
+			// Initial: [t5/10, t6/11]
+			init: entryID{term: 5, index: 10}.append(6).LogSlice,
+			// Append: [t7/12, t7/13, t7/14, t9/15, t9/16, t9/17, t10/18, t10/19, t10/20]
+			app: entryID{term: 6, index: 11}.append(7, 7, 7, 9, 9, 9, 10, 10, 10).LogSlice,
+			// Test case: [t6/11, t7/12, t7/13, t7/14, t9/15, t9/16, t9/17, t10/18, t10/19, t10/20]
+			want: entryID{term: 6, index: 10}.append(6, 7, 7, 7, 9, 9, 9, 10, 10, 10).LogSlice,
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			// Initialize cache with the first entry of init
+			cache := NewTermCache(tc.cacheSize, tc.init.prev)
+			cache.checkInvariants(t)
+			// First append to establish initial state
+			cache.truncateAndAppend(tc.init)
+			cache.checkInvariants(t)
+			// Second append with new entries
+			cache.truncateAndAppend(tc.app)
+			cache.checkInvariants(t)
+
+			func(t *testing.T) {
+				_, found := cache.term(tc.want.prev.index)
+				require.False(t, found, "Looking up for term below the cache should result in not found")
+
+				for _, expTermAtIdx := range tc.want.entries {
+					actualTerm, _ := cache.term(expTermAtIdx.Index)
+					require.Equal(t, expTermAtIdx.Term, actualTerm, "unexpected term for index %d", expTermAtIdx.Index)
+				}
+			}(t)
+		})
+	}
+}
+
+// TestTermCacheRandomInsertMemorySliceCapacity inserts randomly generated
+// logSlices and test that the cache capacity never grows to more than 2x initial size
+func TestTermCacheRandomInsertMemorySliceCapacity(t *testing.T) {
```

Comment:
IMO this test is unnecessary in the way it's framed. We can check this kind of capacity assertion in `checkInvariants`, and all tests including `TestTermCachePopulatedTruncateAndAppend` will cover for it.

The only place where the slice is appended to is `truncateAndAppend` method, and it maintains its cap pretty trivially. The "cost" of maintaining this test is thus higher than the benefit.

But you could remove unnecessary things from it, and leave as a simple randomized test that ensures no invariants are broken under longer sequences of appends than covered by other tests.